Intent is to avoid having to re-process 401 challenge with each request.

kghmanuel opened the pull request.

Commit:
Intent is to avoid having to re-process 401 challenge with each request.

gmarintes reviewed (approved):
code looks okay.
might want to fix the indent on HttpProxy.java line 101.

Commit:
tabs to spaces

rjrudin commented on src/main/java/com/marklogic/spring/security/authentication/MarkLogicAuthenticationManager.java:
```diff
@@ -24,98 +37,129 @@
+        this.pathToAuthenticateAgainst = pathToAuthenticateAgainst;
+    }
+    
+    private RestTemplate prepareRestTemplate(SimpleCredentialsProvider provider) {
```
I think this needs to be protected in case a client e.g. wants to subclass this and override how the CloseableHttpClient is constructed.

rjrudin commented on src/main/java/com/marklogic/spring/security/authentication/MarkLogicAuthenticationManager.java:
```diff
@@ -24,98 +37,129 @@
+        // Create AuthCache instance
+        final AuthCache authCache = new BasicAuthCache();
+        // Generate DIGEST scheme object, initialize it and add it to the local digest cache
+        final DigestScheme digestAuth = new DigestScheme();
```
Does this work for basic auth too?

kghmanuel replied:
no. it does not. fixed.

rjrudin commented on src/main/java/com/marklogic/spring/http/proxy/HttpProxy.java:
```diff
@@ -91,10 +97,45 @@ public void proxy(String path, HttpServletRequest httpRequest, HttpServletRespon
         return HttpMethod.valueOf(request.getMethod());
     }
+    
+    private RestOperations getSessionRestTemplate(HttpServletRequest httpRequest) {
```
Should be protected in case a client wants to subclass this and override the behavior

rjrudin commented on src/main/java/com/marklogic/spring/http/proxy/HttpProxy.java:
```diff
@@ -4,24 +4,30 @@
  * Simple proxy class that uses Spring's RestOperations to proxy servlet requests to MarkLogic.
  */
 public class HttpProxy extends RestClient {
+    public static final String RESTTEMPLATE_SESSION_KEY = "upstream.restemplate";
```
I think something marklogic-specific would be good here to guarantee uniqueness - e.g. "com.marklogic.spring.http.proxy.RestTemplate"

rjrudin commented on src/main/java/com/marklogic/spring/http/proxy/HttpProxy.java:
```diff
@@ -91,10 +97,45 @@ public void proxy(String path, HttpServletRequest httpRequest, HttpServletRespon
+        HttpSession downstreamSession = httpRequest.getSession();
+        if (downstreamSession == null) {
+            //this shouldn't happen.
+            return null;
```
Shouldn't this throw an error now, or else an expected NullPointerException will be thrown later?

rjrudin commented on src/main/java/com/marklogic/spring/http/proxy/HttpProxy.java:
```diff
@@ -91,10 +97,45 @@ public void proxy(String path, HttpServletRequest httpRequest, HttpServletRespon
         }
         HttpMethod method = determineMethod(httpRequest);
-        return getRestOperations().execute(uri, method, requestCallback, responseExtractor);
+        RestOperations client = getSessionRestTemplate(httpRequest);
```
I am wondering if a boolean would be useful to turn this behavior off - it can be on by default, but in case it doesn't work for someone, a boolean to flip it off would be useful - e.g. "private boolean storeRestTemplatesInSession = true" with a getter/setter.

Commit:
1. auto-detect realm
2. adjustments as per PR review
3. as discussed with rqoue, there is no direct need to pass resttemplate from token to session.

Commit:
1. no need to touch auth manager, everything can sit on httpproxy as extension of restclient
2. have cacheing as configurable.

Commit:
fixed whitespace diffs

kghmanuel commented:
made several adjustments to simplify the changes. addressed all concerns.

rjrudin commented on src/main/java/com/marklogic/spring/http/RestConfig.java:
```diff
@@ -7,4 +7,6 @@
 
     public String getScheme();
+
+    public Boolean getCacheDigest();
```
A boolean method normally starts with "is" or "has". Looking at how it's used, perhaps it should be named "isDigestCachingEnabled" or "shouldCacheDigest". I think I prefer the former.

rjrudin commented on src/main/java/com/marklogic/spring/http/proxy/HttpProxy.java:
```diff
@@ -23,19 +31,26 @@
-        return new RestTemplate(new HttpComponentsClientHttpRequestFactory(client));
+        AuthenticationHeader challenge = AuthenticationHeader.getOption(getRestConfig());
+        RestTemplate result = null;
+        if ("digest".equalsIgnoreCase(challenge.getType()) && getRestConfig().getCacheDigest()) {
```
It looks like nothing needs to be done if "getCacheDigest" returns false. So I think the first "if" is to test that, and if it's false, then just return the normal RestTemplate. Else, determine if digest is being used and if so, call prepareDigestTemplate.

kghmanuel replied:
i had comments before that prefers to have a single return point. will adjust.

rjrudin commented on src/main/java/com/marklogic/spring/http/proxy/HttpProxy.java:
```diff
@@ -97,4 +112,34 @@ public void proxy(String path, HttpServletRequest httpRequest, HttpServletRespon
         return HttpMethod.valueOf(request.getMethod());
     }
+    
+    protected RestTemplate prepareDigestTemplate(CredentialsProvider provider, String realm) {
```
Looking at this and the newRestTemplate method in RestClient, I think a RestTemplateBuilder interface is warranted, and this logic is non-trivial. Something like this:

RestTemplate buildRestTemplate(RestConfig, CredentialsProvider);

That makes it easy to modify this logic in case anything is wrong with either implementation.

kghmanuel replied:
In such a scenario where the implementation is faulty, user already has the flexibility to override the function.

rjrudin commented on src/main/java/com/marklogic/spring/http/proxy/HttpProxy.java:
```diff
@@ -97,4 +112,34 @@ public void proxy(String path, HttpServletRequest httpRequest, HttpServletRespon
+                HttpClientBuilder
+                    .create()
+                    .setDefaultCredentialsProvider(provider)
+                    .useSystemProperties()
```
I noticed this was not in the RestClient.newRestTemplate method, should it be?

rjrudin commented on src/main/java/com/marklogic/spring/http/AuthenticationHeader.java:
```diff
@@ -0,0 +1,58 @@
+import org.apache.http.client.methods.HttpHead;
+import org.apache.http.impl.client.HttpClientBuilder;
+
+public class AuthenticationHeader {
```
It seems like the purpose of this class is to extract the digest challenge string and the realm. I think an interface is warranted for this as well, as the logic for doing so isn't trivial. 

How about - AuthenticationHeaderExtractor, with a method of:

AuthenticationHeader extractAuthenticationHeader(RestConfig)

I get nervous when I see a static method for a "plumbing" library like marklogic-spring-web, because that makes it harder for a client to override the functionality. But by defining an interface - which I believe the RestTemplateBuilder below would make use of - it's much easier for a client to customize how the plumbing works.

Commit:
as per code review

rjrudin commented:
@kghmanuel Sorry for the delay, I had a chance to look at the code locally. Take a look at the PR I put together based on your work - https://github.com/rjrudin/marklogic-spring-web/pull/2 

The main change was removing construction of the RestOperations from RestClient/HttpProxy and putting it into DefaultRestOperationsFactory. That keeps RestClient unaware of how a RestOperations is constructed. 

All of your code is in DefaultRestOperationsFactory (DROF). Now, if someone needs to tweak it, they can either extend DROF or write their own code for controlling how a RestOperations is constructed. 

I'm going to close this one, and please let me know if the way the code is organized in that PR works for you.
